[develop2] knowledge base (#2910)
conan-io/docs: knowledge/guidelines.rst

Core guidelines
===============

Good practices
--------------

> - The recipes' ``generate()`` method purpose is to prepare the build as much as possible.

Reviewer: Maybe we can reference the [tutorial part for the generate](https://docs.conan.io/en/2.0/tutorial/creating_packages/preparing_the_build.html) here? Also, I think we can add some notes in each section, like a note of good practices for each part of the documentation. But we can do that in the future starting from this guide

Author: I don't want to distract too much from this guidelines, driving users to different places in the docs, but this is like a "checklist" to go through, making sure you are following them. If you need to understand what the ``generate()`` method is, maybe it is too early to understand the "good practices"?
(This is a bit extreme opinion, I am fine with having some links, as long as it is not cluttered with many of them, that it appears more like an index to continue navigation than a checklist)

I have already started to add them in different places in the docs, like https://docs.conan.io/en/2.0/reference/commands/cache.html (scroll to the bottom)

Reviewer 2: > but this is like a "checklist" to go through, making sure you are following them. If you need to understand what the generate() method is, maybe it is too early to understand the "good practices"?

This idea and belief should be the first sentence on the page so people know what to expect and set them up for success 🙏

  Users calling ``conan install`` will execute this method, and the generated files should
  allow users to do "native" builds (calling directly "cmake", "meson", etc.) as easy as possible.
  Thus, avoiding as much as possible any logic in the ``build()`` method, and moving it to
  the ``generate()`` method helps developers achieve the same build locally as the one that 
  would be produced by a ``conan create`` build in the local cache.
- Always use your own profiles in production, instead of relying on the auto-detected profile,
  as the output of such auto detection can vary over time, resulting in unexpected results.
  Profiles (and many other configuration), can be managed with ``conan config install``.
- Developers should not be able to upload to "development" and "production" repositories in the server.
  Only CI builds have write permissions in the server. Developers should only have read permissions and 
  at most to some "playground" repositories used to work and share things with colleagues,
  but which packages are never used, moved or copied to the development or production repositories.

> - The ``test_package`` purpose is to check that the package has been correctly created (that is, 

Reviewer: It would be nice to add a link here as well: https://docs.conan.io/en/2.0/tutorial/creating_packages/test_conan_packages.html

Core guidelines
===============

This section summarizes some of the good and bad practices.

> Good  and bad practices

Reviewer 3: ```suggestion
Good and bad practices
```

> This section summarizes some of the good and bad practices.
> 
> Good and bad practices
> -----------------------

Reviewer: Maybe saying good and bad is a bit redundant? 

```suggestion

Good practices
--------------
```

> - The recipes ``generate()`` method purpose is to prepare the build as much as possible.

Reviewer 3: ```suggestion
- The recipes' ``generate()`` method purpose is to prepare the build as much as possible.
```

  Users calling ``conan install`` will execute this method, and the generated files should
  allow users to do "native" builds (calling directly "cmake", "meson", etc.) as easy as possible.
  Thus, avoiding as much as possible any logic in the ``build()`` method, and moving it to
  the ``generate()`` method helps developers achieve the same build locally as the one that 
  would be produced by a ``conan create`` build in the local cache.

> - Use always your own profiles in production, instead of relying on the auto-detected profile,

Reviewer 3: ```suggestion
- Always use your own profiles in production, instead of relying on the auto-detected profile,
```

  as the output of such auto detection can vary over time, resulting in unexpected results.
  Profiles (and many other configuration), can be managed with ``conan config install``.

> - Developers should not be able to upload to "development" and "production" repositories in the server. only
>   CI builds has write permissions in the server. Developers should only have read permissions and 

Reviewer 3: ```suggestion
- Developers should not be able to upload to "development" and "production" repositories in the server.
  Only CI builds have write permissions in the server. Developers should only have read permissions and 
```

>   at most to some "playground" repositories used to work, share things with colleagues, but that
>   never used, moved, copied to development or production repositories.

Reviewer 3: ```suggestion
  at most to some "playground" repositories used to work and share things with colleagues,
  but which packages are never used, moved or copied to the development or production repositories.
```

- The ``test_package`` purpose is to check that the package has been correctly created (that is, 
  that it has correctly packaged the headers, the libraries, etc, in the right folders), not that
  the functionality of the package is correct. Then, it should be kept as simple as possible, like
  building and running an executable that uses the headers and links against a packaged library

>   should be enough. Such execution should be as simple as possible too. To run any kind of 
>   unit and functional tests, that should be done in the ``build()`` method.

Reviewer 3: ```suggestion
  should be enough. Such execution should be as simple as possible too. Any kind of 
  unit and functional tests should be done in the ``build()`` method.
```
